clone/cmd/ctclone/ctclone.go

```diff
@@ -0,0 +1,135 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ctclone is a one-shot tool for downloading entries from a CT log.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/clone/internal/download"
+	"github.com/google/trillian-examples/clone/internal/logdb"
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const getEntriesFormat = "/ct/v1/get-entries?start=%d&end=%d"
+
+var (
+	logURL         = flag.String("log_url", "", "Log storage root URL, https://ct.googleapis.com/rocketeer")
+	mysqlURI       = flag.String("mysql_uri", "", "URL of a MySQL database to clone the log into. The DB should contain only one log.")
+	fetchBatchSize = flag.Uint("fetch_batch_size", 32, "The number of entries to fetch from the log in each request.")
+	writeBatchSize = flag.Uint("write_batch_size", 32, "The number of leaves to write in each DB transaction.")
+	workers        = flag.Uint("workers", 2, "The number of worker threads to run in parallel to fetch entries.")
+)
+
+type getEntriesResponse struct {
+	Leaves []json.RawMessage `json:"entries"`
+}
+
+func main() {
+	flag.Parse()
+
+	if len(*logURL) == 0 {
+		glog.Exitf("Missing required parameter 'log_url'")
+	}
+	if len(*mysqlURI) == 0 {
+		glog.Exitf("Missing required parameter 'mysql_uri'")
+	}
+
+	ctx := context.Background()
+
+	db, err := logdb.NewDatabase(*mysqlURI)
+	if err != nil {
+		glog.Exitf("Failed to connect to database: %q", err)
+	}
+	// TODO(mhutchinson): Persist log URI and check here to prevent user-error writing different logs to same DB.
+	head, err := db.Head()
+	if err != nil {
+		if err == logdb.ErrNoDataFound {
+			glog.Infof("failed to find head of database, assuming empty and starting from scratch: %v", err)
+			head = -1
+		} else {
+			glog.Exitf("Failed to query for head of local log: %q", err)
+		}
+	}
+	next := uint64(head + 1)
+
+	fetcher := download.NewHTTPFetcher(*logURL)
+
+	errChan := make(chan error)
+	lc := make(chan []byte, *writeBatchSize*2)
+	go download.Bulk(next, certLeafFetcher{fetcher}, *workers, *fetchBatchSize, lc, errChan)
+
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	lastReported := next
+
+	commit := make([][]byte, *writeBatchSize)
+	ci := 0
+	cs := next
+
+	for {
+		select {
+		case err := <-errChan:
+			glog.Exit(err)
+		case <-ticker.C:
+			delta := next - lastReported
+			rate := float64(delta) / 5
+			glog.Infof("%.1f leaves/s, next leaf=%d", rate, next)
+			lastReported = next
+		case l := <-lc:
+			next++
+			commit[ci] = l
+			ci = (ci + 1) % int(*writeBatchSize)
+			if ci == 0 {
+				db.WriteLeaves(ctx, cs, commit)
```

Comment:
Should this be checking the error returned by `WriteLeaves`, or if this is not necessary can you add a comment outlining why?